### PR TITLE
Switching code that can use FunctionOpInterface instead of func::FuncOp to use it.

### PR DESCRIPTION
**Pull request by benvanik**
There's lots of tech debt around func.func, it turns out. As many uses of func (besides creation) have been switched to the FunctionOpInterface as a first step toward support non-func dialects. These changes will also allow more passes to correctly operate on initializers and future dispatch export ops that carry more information than func.func allows. Codegen is welcome to keep using func where it wants as it's more aligned with upstream than the rest of the compiler but there was a lot of unnecessary coupling to func that predated the existence of the FunctionOpInterface/CallOpInterface this cleans up.

There's a few places that still rely on func that don't really need it but they aren't trivial to switch and the hope is that the bulk of the cleanup here (passes, basic walks/iteration, etc) will make them easier to burn down now that they are the exception rather than the rule.

Progress on #16187.

### CODE REVIEW

File: compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/test/canonicalization.mlir [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt --iree-stablehlo-canonicalize --allow-unregistered-dialect --split-input-file %s | FileCheck %s
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-stablehlo-canonicalize))" --allow-unregistered-dialect --split-input-file %s | FileCheck %s
```
Comment (ScottTodd) [inline compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/test/canonicalization.mlir:1]: I much prefer when tests can omit `--pass-pipeline`, since I can never remember off the top of my head what syntax to use like here. Oh well :/

File: compiler/src/iree/compiler/Bindings/Native/Transforms/Passes.cpp [hunk 2]
```diff
@@ -8,13 +8,18 @@
...
+#include "iree/compiler/Utils/PassUtils.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
 #include "mlir/Pass/PassOptions.h"
 #include "mlir/Pass/PassRegistry.h"
 #include "mlir/Transforms/Passes.h"
 
 namespace mlir::iree_compiler::IREE::ABI {
 
+using FunctionLikeNest =
+    MultiOpNest<func::FuncOp, IREE::Util::InitializerOp, IREE::Util::FuncOp>;
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Bindings/Native/Transforms/Passes.cpp:21]: Could this be written to use FunctionOpInterface instead of an explicit enumeration?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/PromoteTensorLoads.cpp
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/PromoteTensorLoads.cpp:1]: Deleted this entire file? Intentional? Maybe that should go in a standalone PR?

File: docs/website/docs/robots.txt
Comment (ScottTodd) [inline docs/website/docs/robots.txt:1]: wat? why was this file touched? revert?

File: compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/test/canonicalization.mlir (hunk 1 above)
Comment (benvanik) [reply compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/test/canonicalization.mlir:1]: the inconsistency has always bothered me - now at least we're consistently on the most featureful form

File: compiler/src/iree/compiler/Bindings/Native/Transforms/Passes.cpp (hunk 2 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Bindings/Native/Transforms/Passes.cpp:21]: unfortunately not yet - the pass manager can't nest on interfaces so if you add the pass it can't parallelize and will run at the module scope serially ;(

File: docs/website/docs/robots.txt
Comment (benvanik) [reply docs/website/docs/robots.txt:1]: was checked in with the wrong line ending metadata and every time I run bazel-to-cmake it messes it up - just letting it ride here

Comment (ScottTodd) [reply docs/website/docs/robots.txt:1]: Why is bazel-to-cmake editing robots.txt?? 👀

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/PromoteTensorLoads.cpp
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/PromoteTensorLoads.cpp:1]: it was not even being built and has bitrotted - someone removed it from builds but didn't delete the file - this fixes the glitch
image

**Review by ScottTodd — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt --iree-codegen-cpu-materialize-encoding --canonicalize --cse --split-input-file %s | FileCheck %s
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-codegen-cpu-materialize-encoding),canonicalize,cse)" --split-input-file %s | FileCheck %s
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/CPU/test/llvmcpu_materialize_encoding.mlir:1]: This is kind of unfortunate. It's nice to not have to write out the full pass pipeline syntax, but I guess whatever does the automatic nesting isn't smart enough. Does this work better with `MultiOpNest`?

**Review by qedawkins — APPROVED** [review]
200w

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/reduce_bank_conflicts.mlir [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt %s --split-input-file --iree-codegen-gpu-reduce-bank-conflicts  | FileCheck %s
+// RUN: iree-opt %s --split-input-file --pass-pipeline="builtin.module(func.func(iree-codegen-gpu-reduce-bank-conflicts))" | FileCheck %s
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/reduce_bank_conflicts.mlir:1]: This kind of sucks. I wish the pipeline would just apply to `func.func` even though its through the interface. I guess it can get tricky with nested ops implementing the interface.

**Review by MaheshRavishankar — APPROVED** [review]
THis is some sweep! Thanks!

**Comment by ScottTodd** [comment]
Nice to see I'm not alone in disliking `--pass-pipeline="builtin.module(func.func(...` in tests :P

**Comment by qedawkins** [comment]
oh I didn't notice you commented about it too, yeah I think it's also a problem when trying to run one-off passes outside of tests (scripting/debugging). I bet if we go fix a bunch of passes upstream to use the interface support gets added pretty soon after :)

**Comment by benvanik** [comment]
contributions to make upstream better very welcome :)
